File: cpp/programmers/네트워크.cpp
#include <string>
#include <vector>

using namespace std;

// v1
// void dfs(int i, int n, int (&visited)[200], vector<vector<int>>& computers){
//     vector<int> stack;
//     stack.push_back(i);
//     visited[i] = 1;

//     while (!stack.empty()){
//         int i = stack.back();
//         stack.pop_back();
//         for (int j=0; j<computers.size(); ++j){
//             if (computers[i][j] == 1 && visited[j] == 0){
//                 stack.push_back(j);
//                 visited[j] = 1;
//             }
//         }
//     }
// }


// int solution(int n, vector<vector<int>> computers) {
//     int answer = 0;
//     int visited[200] = {0};
    
    
//     for (int i=0; i<n; ++i){
//         if (visited[i] == 0){
//             dfs(i, n,visited, computers);
//             ++answer;
//         }
//     }
    
//     return answer;
// }


// v2
void dfs(int i, int n, vector<int>& visited, vector<vector<int>>& computers){
    vector<int> stack;
    stack.push_back(i);
    visited[i] = 1;

    while (!stack.empty()){
        int i = stack.back();
        stack.pop_back();
        for (int j=0; j<computers.size(); ++j){
            if (computers[i][j] == 1 && visited[j] == 0){
                stack.push_back(j);
                visited[j] = 1;
            }
        }
    }
}


int solution(int n, vector<vector<int>> computers) {
    int answer = 0;
    vector<int> visited(n,0);
    
    for (int i=0; i<n; ++i){
        if (visited[i] == 0){
            dfs(i, n, visited, computers);
            ++answer;
        }
    }
    
    return answer;
}

File: python/programmers/네트워크.py

# v1
# def dfs(i):
#     stack = []
#     visited[i] = 1
#     stack.append(i)
#
#     while stack:
#         i = stack.pop()
#         idx = 0
#         for j in computers[i]:
#             if visited[idx] == 0 and j == 1:
#                 stack.append(idx)
#                 visited[idx] = 1
#             idx += 1
#
# def solution(n, computers):
#     global visited
#     visited = [0]*n
#     cnt = 0
#     for i in range(n):
#         if visited[i] == 0:
#             dfs(i)
#             cnt += 1
#
#     answer = cnt
#     return answer


# v2


def solution(n, computers):
    visited = [0]*n
    def dfs(i):
        stack = []
        visited[i] = 1
        stack.append(i)

        while stack:
            i = stack.pop()
            for j in range(len(computers[i])):
                if visited[j] == 0 and computers[i][j] == 1:
                    stack.append(j)
                    visited[j] = 1
    cnt = 0
    for i in range(n):
        if visited[i] == 0:
            dfs(i)
            cnt += 1

    answer = cnt
    return answer
n = int(input())
computers = [list(map(int, input().split(','))) for _ in range(n)]
print(solution(n, computers))